### PR TITLE
Main.js config: Fix Builder type

### PR DESCRIPTION
**Pull request by tyankatsu0105**
Issue:

## What I did
In this PR, Builder type is improved.

https://www.typescriptlang.org/play?#code/C4TwDgpgBA8gNgEwMIHsBO0C8UDeAoKQqAIwFcBLRCNALigHIB3CYsAQwGMBrAFnqgA+DZq05cArPyEBnYGnIA7AOZ4AvnjwcUC2VBSJUGOvGTosuAkTKUE1OvXpqNoSFAByERofP4iJClS0wizs3HyCwaLckhEAFLLyylAAZLiqAJROmtq6Cp7edB5eZlDYvlYBtkH0suggxCgoXAC01oHNAG7kwBCOqkA

Autocomplete is not working...
スクリーンショット 2021-09-11 17 05 04

Yeah!! Autocomplete is working!!
スクリーンショット 2021-09-11 17 05 18

## How to test

- [ ] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [ ] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

**Review by darleendenno — APPROVED** [review]
Thanks @tyankatsu0105 for taking the time to improve Storybook's types! I'm good with this change. @ndelangen are you?

**Comment by darleendenno** [comment]
@tmeasday does this affect what you're doing with the builder updates?

**Comment by tyankatsu0105** [comment]
@darleendenno 
I"m glad to hear that :)
I don't know the solution to some failure CI.
Do you know how to treat these?
I'm willing to try that if you share your knowledge.

**Comment by tmeasday** [comment]
Hmmm, doesn't affect me but I am not sure what `string & {}` means exactly?

**Comment by tyankatsu0105** [comment]
@tmeasday 

Now, we can't input strings other than `webpack4` and `webpack5`. It is caused by [this line](https://github.com/storybookjs/storybook/blob/68e638a09eaad8e543dc2640ebd40894371266b1/lib/cli/src/project_types.ts#L94).

See this PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/55316
Actually, `(string | {})` seems like hacking.
It helps us to input some strings(like `foobar`). And also we can get autocomplete provided by Editor.

**Comment by tmeasday** [comment]
OK, cool, yeah makes sense to me, that link led me to : https://github.com/microsoft/TypeScript/issues/29729

I'm not sure how commonly known this is, perhaps a comment would help some future confused person.

File: lib/cli/src/project_types.ts [hunk 1]
```diff
@@ -91,7 +91,7 @@ export enum CoreBuilder {
   Webpack5 = 'webpack5',
 }
 
-export type Builder = CoreBuilder | string;
+export type Builder = CoreBuilder | (string | {});
```
Comment (tmeasday) [inline lib/cli/src/project_types.ts:94]:
```suggestion
// The `& {}` bit allows for auto-complete, see: https://github.com/microsoft/TypeScript/issues/29729
export type Builder = CoreBuilder | (string & {});
```

**Review by shilman — APPROVED** [review]
Thanks @tyankatsu0105 !! 🙏

**Pull request merged**